src/protocols/http/hpr_http_roaming.erl

```diff
@@ -0,0 +1,515 @@
+]).
+
+%% Tokens
+-export([
+    make_uplink_token/5,
+    parse_uplink_token/1
+]).
+
+-export([new_packet/3]).
+
+-define(NO_ROAMING_AGREEMENT, <<"NoRoamingAgreement">>).
+
+-define(JOIN1_DELAY, 5_000_000).
+-define(JOIN2_DELAY, 6_000_000).
+-define(RX2_DELAY, 2_000_000).
+-define(RX1_DELAY, 1_000_000).
+
+%% Roaming MessageTypes
+-type prstart_req() :: map().
+-type prstart_ans() :: map().
+-type xmitdata_req() :: map().
+-type xmitdata_ans() :: map().
+
+-type netid_num() :: non_neg_integer().
+-type gateway_time() :: non_neg_integer().
+
+-type downlink() :: {
+    ResponseStream :: gateway_stream(),
+    Response :: any()
+}.
+
+-type region() :: atom().
+-type token() :: binary().
+-type dest_url() :: binary().
+-type flow_type() :: sync | async.
+
+-define(TOKEN_SEP, <<"::">>).
+
+-record(packet, {
+    packet_up :: hpr_packet_up:packet(),
+    gateway_time :: gateway_time(),
+    response_stream :: gateway_stream()
+}).
+-type packet() :: #packet{}.
+
+-type gateway_stream() :: pid().
+
+-type downlink_packet() :: hpr_packet_down:packet().
+
+-export_type([
+    netid_num/0,
+    packet/0,
+    gateway_time/0,
+    downlink/0,
+    downlink_packet/0,
+    gateway_stream/0
+]).
+
+%% ------------------------------------------------------------------
+%% Uplink
+%% ------------------------------------------------------------------
+
+-spec new_packet(
+    PacketUp :: hpr_packet_up:packet(),
+    GatewayTime :: gateway_time(),
+    GatewayStream :: gateway_stream()
+) -> #packet{}.
+new_packet(PacketUp, GatewayTime, GatewayStream) ->
+    #packet{
+        packet_up = PacketUp,
+        gateway_time = GatewayTime,
+        response_stream = GatewayStream
+    }.
+
+-spec make_uplink_payload(
+    NetID :: netid_num(),
+    Uplinks :: list(packet()),
+    TransactionID :: integer(),
+    DedupWindowSize :: non_neg_integer(),
+    Destination :: binary(),
+    FlowType :: sync | async
+) -> prstart_req().
+make_uplink_payload(
+    NetID,
+    Uplinks,
+    TransactionID,
+    DedupWindowSize,
+    Destination,
+    FlowType
+) ->
+    #packet{
+        packet_up = PacketUp,
+        gateway_time = GatewayTime,
+        response_stream = ResponseStream
+    } = select_best(Uplinks),
+    Payload = hpr_packet_up:payload(PacketUp),
+    PacketTime = hpr_packet_up:timestamp(PacketUp),
+
+    PubKeyBin = hpr_packet_up:gateway(PacketUp),
+    Region = hpr_packet_up:region(PacketUp),
+    DataRate = hpr_packet_up:datarate(PacketUp),
+    Frequency = hpr_packet_up:frequency_mhz(PacketUp),
+
+    {RoutingKey, RoutingValue} = routing_key_and_value(PacketUp),
+
+    Token = make_uplink_token(PubKeyBin, Region, PacketTime, Destination, FlowType),
+
+    ok = hpr_http_roaming_utils:insert_handler(PubKeyBin, ResponseStream),
+
+    VersionBase = #{
+        'ProtocolVersion' => <<"1.1">>,
+        'SenderNSID' => <<"">>,
+        'DedupWindowSize' => DedupWindowSize
+    },
+
+    VersionBase#{
+        'SenderID' => <<"0xC00053">>,
```

Comment:
This looks like our hardcoded Net ID, does it matter?

Reply:
@michaeldjeffrey , can you explain?

Reply:
We don't have anything setup under the second NetID. And this is our first "official" NetID. 
Including the source NetID is part of the spec for back and forth reasons.
We could put it in settings file. But we could also wait until it needs to change to do that.

Reply:
Shouldn't it be extracted from the `devaddr`? Because this is the roaming spec and we are supporting multiple networks after all?

Reply:
This field is `SenderID`. 
It's to identify the network that heard the packet.

The NetID encoded in the `devaddr` is what we use to fill out the `ReceiverID`.
For joins, we have to know, which is why NetID is part of a Route.

Reply:
To flesh things out a bit, on the downlink side.

The `ReceiverID` should always be our NetID, and `SenderID` is the NetID who is asking us to downlink.

Reply:
OK make sense thx